rpc/src/server/service.rs

```diff
@@ -406,6 +446,197 @@ mod fns {
+                if let Some(Bucket::Exists(roll)) = data.get(&successor_key) {
+                    next_roll = roll
+                } else {
+                    break;
+                }
+            }
+        }
+        Ok(Some(roll_owners))
+    }
+
+    pub(crate) fn get_baking_rights(block_id: &str, delegate: Option<String>, level: &Option<String>, cycle: &String, max_priority: String, has_all: bool, head_level: i32, timestamp: String, list: ContextList, persistent_storage: &PersistentStorage) -> Result<Option<Vec<BakingRights>>, failure::Error> {
+        let mut baking_rights = Vec::<BakingRights>::new();
+        
+        // level of the block specified in the url params
+        let block_level = match get_block_level(block_id, persistent_storage, &list).unwrap() {
+            Some(v) => v,
+            None => panic!("Cannot get level")  // Cannot get level
+        };
+
+        let requested_cycle = match cycle.parse() {
+            // cycle requested from query
+            Ok(cycle) => {
+                // check whether the cycle is reachable
+                // the storage stores cycle up to: current_cycle + perserved_cycles(5) + 2
+                // e.g. if head is in cycle 66, we can get the rights for up to cycle 73
+                if ((cycle as i32 - (block_level as i32 / 2048)).abs() as i32) <= 7 {
+                    cycle
+                } else {
+                    panic!("Cycle not found in block: {}", block_id)
+                }
+            },
+            // no cycle requested
+            Err(_e) => -1
+        };
+
+        // If no level is specified, we return the next block to be baked
+        let mut requested_level = match level {
+            Some(level) => level.parse().unwrap(),
+            None => head_level + 1,
+        };
+
+        // if a cycle is specified, we need to iterate through all of the levels in the cycle
+        let level_to_itarate_to = if requested_cycle >= 0 {
+            requested_level = requested_cycle * 2048 as i32; // first block of the cycle
+            requested_level + 2048
+        } else {
+            requested_level
+        };
+
+        // get the requested delegate from the query, set to all if not specified
+        let requested_delegate = match delegate {
+            Some(delegate) => delegate,
+            None => "all".to_string() // all
+        };
+        // random number generator
+        let mut rng = rand::thread_rng();
+
+        // get the rolls from the context storage
+        let rolls = get_rolls(block_id, &cycle, persistent_storage, list).unwrap();
+        let roll_owners = match rolls {
+            Some(r) => r,
+            None => panic!("Error getting rolls")
+        };
+
+        // iterate through the whole cycle if necessery
+        for level in requested_level..(level_to_itarate_to + 1) {
+            // 30 is from the protocol constants
+            let seconds_to_add = (level - head_level).abs() * 30;
```

Comment:
It would be probably a good idea to actually read the value from the context constants. There is already prepared encoding for all protocols to decode them from context